### DISCUSSION TITLE
How to modify duty cycle or frequency or period in timer callback

### DISCUSSION THREAD

**Discussion by mogplus88** [micropython/micropython#16281] [ESP32] [answered]
I must have missed something somewhere. How can the mark/space ratio of a PWM stream be set in an ESP32 timer? All I can see is freq and period, which are nearly synonyms for each other anyway. If I wanted to fade a led or set led brightness with an ESP32 timer, how would I do it?

Thanks.

**Comment by sosi-deadeye** [comment]
The `init` method can reconfigure a timer. The timer calls the callback and the first argument is the timer itself.

Example:
```python
from machine import Timer


def handler1(timer):
    print("[handler1]: Function triggered by timer", timer)
    timer.init(callback=handler2, freq=1)

def handler2(timer):
    print("[handler2]: Function triggered by timer", timer)


t1 = Timer(1)
t1.init(mode=Timer.PERIODIC, freq=1, callback=handler1)
```
docs: https://docs.micropython.org/en/latest/library/machine.Timer.html#class-timer-control-hardware-timers

**Reply by mogplus88** [reply]
@sosi-deadeye Thanks for the quick reply.
So how to I modify the mark/space ratio? If I wanted the pin to be high for 2000us and low for 300us for instance? Or a 20% duty cycle if I wanted to dim a led?
Thanks.

**Reply by robert-hh** [reply]
You will need both a PWM object and a Timer. The PWM is set to create a PWM signal at the PIN with the LED. It will be set to a certain frequency, like 1000Hz using pwm.freq(n). The brightness of the LED can then be changed by using pwm.duty_u16(n). n is in the range of 0 to 65536. 
The timer will be used to change the duty cycle of the PWM as needed in the timer callback.

**Comment by sosi-deadeye** [comment, marked as answer]
The `PWM` object also has the `init` method: https://docs.micropython.org/en/latest/library/machine.PWM.html#machine.PWM.init
In the handler/callback you can change the configuration of the PWM instance.
You could set frequency and duty cycle to get high for 2000 µs and low for 300 µs. Timers are using milliseconds.

Code:
```python
from machine import Pin, PWM

ON_TIME_US = 2_000
OFF_TIME_US = 300
FREQ = int(1_000_000 / (ON_TIME_US + OFF_TIME_US))
DUTY = int((OFF_TIME_US / ON_TIME_US) * 65535)

pwm1 = PWM(Pin(16), freq=FREQ, duty_u16=DUTY)
```
Result:
![grafik](https://github.com/user-attachments/assets/114cb297-862d-4f26-98d4-7608391dfd11)

**Reply by mogplus88** [reply]
@sosi-deadeye Thanks! That looks like exactly what I was looking for!

:-)
